components/script/dom/formdata.rs

```diff
@@ -156,11 +158,35 @@ impl FormData {
     }
 
     pub fn datums(&self) -> Vec<FormDatum> {
-        let mut ret = vec![];
-        for values in self.data.borrow().values() {
-            ret.append(&mut values.clone());
+        self.data.borrow().values()
+            .flat_map(|value| value.iter())
+            .map(|value| value.clone())
+            .collect()
+    }
+
+    pub fn atoms(&self) -> Vec<Atom> {
+        self.data.borrow().iter()
+            .flat_map(|(key, value)| iter::repeat(key.clone()).take(value.len()))
+            .collect()
+    }
+}
+
+impl Iterable for FormData {
+    type Key = USVString;
+    type Value = FileOrUSVString;
+
+    fn get_iterable_length(&self) -> u32 {
+        self.datums().len() as u32
+    }
+
+    fn get_value_at_index(&self, n: u32) -> FileOrUSVString {
+        match self.datums()[n as usize].value {
```

Comment:
And this could be

```rust
let data = self.data.borrow();
let value = &data.values()
                 .flat_map(|value| value.iter())
                 .nth(n as usize)
                 .value;
match value {
```